### PR TITLE
Component serialization

### PR DESCRIPTION
**Pull request by YahimaB**

### CODE REVIEW

File: Engine/BaseAssets/Components/Lights/DirectionalLight.cs [hunk 1]
```diff
@@ -9,19 +9,13 @@ namespace Engine.BaseAssets.Components
...
-            get => shadowSize;
-            set
-            {
-                if (value <= 0)
-                    throw new ArgumentOutOfRangeException("ShadowSize", "Shadow size must be a positive value.");
-                shadowSize = value;
-            }
-        }
+
+        public Ranged<int> ShadowSize => new Ranged<int>(ref shadowSize, 0);
```
Comment (Notrum666) [inline Engine/BaseAssets/Components/Lights/DirectionalLight.cs:15]: ShadowSize must be a positive value

File: Engine/BaseAssets/Components/Lights/SpotLight.cs [hunk 2]
```diff
@@ -9,76 +9,40 @@ namespace Engine.BaseAssets.Components
...
-                shadowSize = value;
-            }
-        }
-        public Matrix4x4f lightSpace
+
+        public Ranged<float> Radius => new Ranged<float>(ref radius, 0.0f);
+        public Ranged<float> Intensity => new Ranged<float>(ref intensity, 0.0f, 1.0f);
+        public Ranged<float> AngularIntensity => new Ranged<float>(ref angularIntensity, 0.0f, 1.0f);
+        public Ranged<float> Angle => new Ranged<float>(ref angle, 0.0f, MathF.PI);
+        public Ranged<int> ShadowSize => new Ranged<int>(ref shadowSize, 0);
```
Comment (Notrum666) [inline Engine/BaseAssets/Components/Lights/SpotLight.cs:29]: ShadowSize must be a positive value

File: Engine/BaseAssets/Components/MeshComponent.cs [hunk 3]
```diff
@@ -1,26 +1,43 @@
...
             }
         }
-        public Material[] Materials { get; private set; } = new Material[0];
+
+        public Material[] Materials => materials;
+
+        public override void OnInspectorFieldChanged(FieldInfo fieldInfo)
+        {
+            if (fieldInfo.Name == nameof(model))
+            {
```
Comment (Notrum666) [inline Engine/BaseAssets/Components/MeshComponent.cs:29]: please remove curly brackets to follow projects codestyle

File: Engine/BaseAssets/Components/SoundListener.cs [hunk 4]
```diff
@@ -13,18 +12,23 @@ public bool IsCurrent
...
                 }
             }
         }
 
-        public SoundListener()
+        protected override void OnInitialized()
         {
             Listener = new Listener();
+
+            if (SoundCore.CurrentListener == null) //TODO: Unity allows only 1 AudioListener in the scene. Otherwise throws warning
```
Comment (Notrum666) [inline Engine/BaseAssets/Components/SoundListener.cs:28]: we allow multiple existing, but only one active

File: Engine/BaseAssets/Components/SoundListener.cs [hunk 5]
```diff
@@ -13,18 +12,23 @@ public bool IsCurrent
...
+            {
```
Comment (Notrum666) [inline Engine/BaseAssets/Components/SoundListener.cs:29]: please remove curly braces to follow projects codestyle

File: Engine/INotifyFieldChanged.cs [hunk 6]
```diff
@@ -4,6 +4,6 @@ namespace Engine
 {
     public interface INotifyFieldChanged
     {
-        public abstract void OnFieldChanged(FieldInfo fieldInfo);
+        public void OnInspectorFieldChanged(FieldInfo fieldInfo);
```
Comment (Notrum666) [inline Engine/INotifyFieldChanged.cs:7]: Nothing related to editor should exist in engine, rename back to OnFieldChanged

**Review by Notrum666 — CHANGES_REQUESTED** [review]

File: Engine/BaseAssets/Components/Camera.cs [hunk 7]
```diff
@@ -154,46 +174,44 @@ private void GenerateBuffers(int width, int height)
...
 
-        public void SwapFrameBuffers()
+        internal void SwapFrameBuffers()
         {
             lock (middlebuffer)
             {
-                FrameBuffer tmp = Backbuffer;
-                Backbuffer = middlebuffer;
-                middlebuffer = tmp;
+                (Backbuffer, middlebuffer) = (middlebuffer, Backbuffer);
```
[the remark this answers is not in the document]
Comment (YahimaB) [reply Engine/BaseAssets/Components/Camera.cs:201]: Meh

File: Engine/BaseAssets/Components/Camera.cs [hunk 8]
```diff
@@ -154,46 +174,44 @@ private void GenerateBuffers(int width, int height)
...
         }
 
         public FrameBuffer GetNextFrontBuffer()
         {
             lock (middlebuffer)
             {
-                FrameBuffer tmp = middlebuffer;
-                middlebuffer = frontbuffer;
-                frontbuffer = tmp;
+                (middlebuffer, frontbuffer) = (frontbuffer, middlebuffer);
```
[the remark this answers is not in the document]
Comment (YahimaB) [reply Engine/BaseAssets/Components/Camera.cs:209]: meh

File: Engine/BaseAssets/Components/Lights/DirectionalLight.cs (hunk 1 above)
Comment (YahimaB) [reply Engine/BaseAssets/Components/Lights/DirectionalLight.cs:15]: +

File: Engine/BaseAssets/Components/Lights/SpotLight.cs (hunk 2 above)
Comment (YahimaB) [reply Engine/BaseAssets/Components/Lights/SpotLight.cs:29]: ++

File: Engine/BaseAssets/Components/MeshComponent.cs (hunk 3 above)
Comment (YahimaB) [reply Engine/BaseAssets/Components/MeshComponent.cs:29]: +

File: Engine/BaseAssets/Components/SoundListener.cs (hunk 4 above)
Comment (YahimaB) [reply Engine/BaseAssets/Components/SoundListener.cs:28]: ++

File: Engine/BaseAssets/Components/SoundListener.cs (hunk 5 above)
Comment (YahimaB) [reply Engine/BaseAssets/Components/SoundListener.cs:29]: ++

File: Engine/INotifyFieldChanged.cs (hunk 6 above)
Comment (YahimaB) [reply Engine/INotifyFieldChanged.cs:7]: ++

**Review by Notrum666 — APPROVED** [review]

**Pull request merged**